.jenkins/Jenkinsfile

```diff
@@ -16,524 +16,486 @@
+    checkout scm
+    tool type: 'jdk', name: "jdk_${jdk}_latest"
+    // FIXME python
+    // FIXME cython
+    def statusCode = 0, attempt = 1
+    retry(2) {
+      if (attempt > 1) { sleep(60 * attempt) }
+      def logfile = "${JOB_NAME}_${BUILD_NUMBER}_${test_type}_jdk${jdk}_python_${python}_${cython}_${arch}_attempt_${attempt}.log"
+      try {
+        //("${JOB_NAME}-${test_type}-${BUILD_NUMBER}")
+
+        dir("build") {
+          unstash name: "${arch}_${jdk}"
+        }
+
+        echo "Executing `${test_script} ${test_type} ${split} ${jdk} 2>&1 | tee ${logfile}` on ${NODE_NAME} with jdk=${jdk} python=${python} cython=${cython}"
+        statusCode = sh returnStatus:true, script:"${test_script} ${test_type} ${split} ${jdk} 2>&1 | tee ${logfile}"
+      } finally {
+        attempt = attempt + 1
+        if (0 == statusCode) { 
+            junit testResults: '**/build/test/**/TEST*.xml,**/cqlshlib.xml,**/nosetests.xml', testDataPublishers: [[$class: 'StabilityTestDataPublisher']]
+            generateUnifiedTestReport()
+        } else {
+          toCopy = ""
+        }
+        sh "xz -f *_attempt_*.log"
+        copyToNightlies("*_attempt_*.log.xz,console.log.xz,TESTS-TestSuites.xml.xz,build/test/logs/**,${toCopy}", "${test_type}/${arch}/jdk${jdk}/${python}/cython_${cython}/" + "split_${split}".replace("/", "_"))
+        cleanAgent(test_script)
+        if (0 != statusCode) { 
+          currentBuild.result = 'FAILURE'
+          error("Failed `${test_script} ${test_type} ${split} ${jdk}` with arch=${arch} jdk=${jdk} python=${python} cython=${cython} on ${NODE_NAME}") 
+        }
       }
     }
+    //if (build_result.result != 'SUCCESS') unstable("${build_name} failures")
+    //if (build_result.result == 'ABORTED' && currentBuild.result != 'FAILURE') currentBuild.result='ABORTED'
+    //if (build_result.result == 'FAILURE') currentBuild.result='FAILURE'
+  }
+}
+
+def copyToNightlies(sourceFiles, remoteDirectory='') {
+  def remotePath = remoteDirectory.startsWith("cassandra/") ? "${remoteDirectory}" : "cassandra/${JOB_NAME}/${BUILD_NUMBER}/${remoteDirectory}"
+  def attempt = 1
+  retry(9) {
+    if (attempt > 1) { sleep(60 * attempt) }
+    sshPublisher(
+      continueOnError: true, failOnError: false,
+      publishers: [
+        sshPublisherDesc(
+         configName: "Nightlies",
+         transfers: [ sshTransfer( sourceFiles: sourceFiles, remoteDirectory: remotePath) ]
+         )
+      ])
   }
+  echo "archived to https://nightlies.apache.org/${remotePath}"
+}
+
+def cleanAgent(job_name) {
+  def maxJobHours = 12
+  echo "Cleaning project, processes, docker for '${job_name}' on ${NODE_NAME}…" ;
+  sh """
+      git clean -qxdff -e build/test/jmh-result.json || true;
+      if pgrep -xa docker || pgrep -af "build/docker" || pgrep -af "cassandra-builds/build-scripts" ; then docker system prune --all --force --filter "until=${maxJobHours}h" || true ; else  docker system prune --force --volumes || true ;  fi;
```

Comment:
Do we run the risk of a lot of back to back jobs filling up hosts since we only prune once > 12 hours? Or is the idea that other build_script / jdk combos could still be legitimately running when we go to clean it off?

Reply:
12 hours because docker cache++ can speed things up.
this is how it is today. i expect this to evolve a lot once we're running in k8s and can benchmark and debug better.